var/spack/repos/builtin/packages/trilinos/package.py

```diff
@@ -832,11 +832,18 @@ def define_tpl(trilinos_name, spack_name, have_dep):
                 libgfortran = fc('--print-file-name',
                                  'libgfortran.a',
                                  output=str).strip()
-            # -L<libdir> -lgfortran required for OSX
-            # https://github.com/spack/spack/pull/25823#issuecomment-917231118
-            options.append(
-                define('Trilinos_EXTRA_LINK_FLAGS',
-                       '-L%s/ -lgfortran' % os.path.dirname(libgfortran)))
+            if spec.version < Version('14'):
+                # Trilinos_EXTRA_LINK_FLAGS do not correctly propagate to
+                # Kokkos-Kernels prior to version 14.0
+                options.append(
+                    define('CMAKE_SHARED_LINKER_FLAGS',
+                           '-L%s/ -lgfortran' % os.path.dirname(libgfortran)))
+            else:
+                # -L<libdir> -lgfortran required for OSX
+                # https://github.com/spack/spack/pull/25823#issuecomment-917231118
+                options.append(
+                    define('Trilinos_EXTRA_LINK_FLAGS',
+                           '-L%s/ -lgfortran' % os.path.dirname(libgfortran)))
```

Comment:
Is there a reason we can't use the spack wrapper flags (`flag_handler`) instead? That should cover all cases that call the compilers. I know this code is very old so that might've been just an oversight.

Reply:
Hi @sethrj. This problem comes up when gfortran (not built by spack) is used to build a static library (in this case openblas) and the static library is linked against by another package with shared libraries. There are a few ways to address this:

1.) Setting 
```yaml
compilers:
- compiler:
    spec: apple-clang@13.1.6
    paths:
      cc: /usr/bin/clang
      cxx: /usr/bin/clang++
      f77: /usr/local/bin/gfortran
      fc: /usr/local/bin/gfortran
    flags:
      ldflags: ' -L/usr/local/Cellar/gcc/11.3.0_2/lib/gcc/11/libgfortran.dylib -lgfortran '
```
or
setting 
```yaml
compilers:
- compiler:
    spec: apple-clang@13.1.6
    paths:
      cc: /usr/bin/clang
      cxx: /usr/bin/clang++
      f77: /usr/local/bin/gfortran
      fc: /usr/local/bin/gfortran
    flags:
      ldflags: -lgfortran
    extra_rpaths:
    - /usr/local/Cellar/gcc/11.3.0_2/lib/gcc/11/
```
Since the issue stems from a misconfigured combination of Clang/gfortran, this seems like a reasonable fix (fixing the configuration). All of the following options are meant to handle the misconfigured compilers after the fact, right?

2.) The fix in this PR that uses `CMAKE_SHARED_LINKER_FLAGS` / `Trilinos_EXTRA_LINK_FLAGS`.

3.) Have openblas provide the extra fortran flags needed so that any package upstream of it will inherit the correct flags.

4.) Use the `flag_handler` to do an equivalent to 2

5.) Open for suggestion

Reply:
I think it's not exclusively due to misconfigured libraries: it might just be the build systems not propagating the implicit language flags. If an upstream static library Food that builds a Fortran-based library and a `FooConfig.cmake`, then CMake will use [`CMAKE_<LANG>_IMPLICIT_LINK_LIBRARIES`](https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_IMPLICIT_LINK_LIBRARIES.html) to propagate `-lgfortran` etc. when linking downstream C++-only code against that fortran code. Presumably because of tribits (which doesn't [at least as of 13.2] look for `Config.cmake` files) and possibly inadequate upstream static installations of fortran tools (such as netlib blas/lapack, openblas), Trilinos could fail when using the C++ linker.

I think the best (most Spack-y) option is number 4, because it's a problem with the combination of trilinos and upstream library.

Reply:
Hi @sethrj,  I'm working towards number 4, but handling it in the `flag_handler`, and I've run into an issue. 

```py
fc = Executable(spec["mpi"].mpifc) if ("+mpi" in spec) else Executable(spack_fc)
```
with 

```py
fc = Executable(spec["mpi"].mpifc)
``` 
producing

```out
==> [2022-10-03-09:38:15.241117] Error: AttributeError: 'Spec' object has no attribute 'mpifc'

/Users/pakuber/test_spack_xyce/spack/var/spack/repos/builtin/packages/trilinos/package.py:480, in flag_handler:
        477                flags.append("-Wl,-undefined,dynamic_lookup")
        478            # Fortran lib (assumes clang is built with gfortran!)
        479            if "+fortran" in spec and spec.compiler.name in ["gcc", "clang", "apple-clang"]:
  >>    480                fc = Executable(spec["mpi"].mpifc) if ("+mpi" in spec) else Executable(spack_fc)
        481                libgfortran = fc("--print-file-name", "libgfortran." + dso_suffix, output=str).strip()
        482                # if libgfortran is equal to "libgfortran.<dso_suffix>" then
        483                # print-file-name failed, use static library instead
```
and 
```py
fc = Executable(spack_fc)
```
results in

```out
==> [2022-10-03-11:48:53.405227] Error: ProcessError: Command exited with status 1:
    '/Users/pakuber/test_spack_xyce/spack/lib/spack/env/clang/gfortran' '--print-file-name' 'libgfortran.dylib'
[spack cc] ERROR: Spack compiler must be run from Spack! Input 'SPACK_ENV_PATH' is missing
```
so it seems like at the time of the call to `flag_handler`, the environment is not yet populated.

Reply:
On further inspection it looks like:

https://github.com/spack/spack/blob/develop/lib/spack/spack/build_environment.py#L807-L809

```py
    if need_compiler:
        set_compiler_environment_variables(pkg, env_mods)
        set_wrapper_variables(pkg, env_mods)
```
has an ordering which populates parts of the environment needed by `set_compiler_environment_variables` which calls the `flag_handler`.

lib/spack//spack/build_environment.py